File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/book/Category.java
package ru.kmorozov.library.data.model.book;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by km on 26.12.2016.
 */

@Document
public class Category {

    @Id
    String id;

    @Indexed(unique = true)
    String name;

    @DBRef(lazy = true)
    Set<Category> parents;

    @DBRef(lazy = true)
    Set<Storage> storages;

    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return this.name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Set<Category> getParents() {
        return this.parents;
    }

    public void setParents(Set<Category> parents) {
        this.parents = parents;
    }

    public Iterable<Storage> getStorages() {
        return this.storages;
    }

    public void setStorages(Set<Storage> storages) {
        this.storages = storages;
    }

    public void addStorage(Storage storage) {
        if (null == this.storages)
            this.storages = new HashSet<>();

        this.storages.add(storage);
    }

    public void addParents(Collection<Category> items) {
        if (null == this.parents)
            this.parents = new HashSet<>();

        this.parents.addAll(items);
    }

    public void addParent(Category parent) {
        if (null == this.parents)
            this.parents = new HashSet<>();

        this.parents.add(parent);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (null == o || this.getClass() != o.getClass()) return false;

        Category category = (Category) o;

        return this.id.equals(category.id);

    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }
}

File: ru.kmorozov.library.data.model/src/main/java/ru/kmorozov/library/data/model/book/Category.kt
package ru.kmorozov.library.data.model.book

import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document
import java.util.HashSet

/**
 * Created by km on 26.12.2016.
 */

@Document
class Category {

    @Id
    lateinit var id: String

    @Indexed(unique = true)
    lateinit var name: String

    @DBRef(lazy = true)
    internal var parents: MutableSet<Category>? = null

    @DBRef(lazy = true)
    internal var storages: MutableSet<Storage>? = null

    fun getParents(): Set<Category>? {
        return parents
    }

    fun setParents(parents: MutableSet<Category>) {
        this.parents = parents
    }

    fun getStorages(): Iterable<Storage>? {
        return storages
    }

    fun setStorages(storages: MutableSet<Storage>) {
        this.storages = storages
    }

    fun addStorage(storage: Storage) {
        if (null == storages)
            storages = HashSet()

        storages!!.add(storage)
    }

    fun addParents(items: Collection<Category>) {
        if (null == parents)
            parents = HashSet()

        parents!!.addAll(items)
    }

    fun addParent(parent: Category) {
        if (null == parents)
            parents = HashSet()

        parents!!.add(parent)
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (null == o || javaClass != o.javaClass) return false

        val category = o as Category?

        return id == category!!.id

    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}
